File: leetcode/medium/substr_three_chars.py
def num_substr(s):
    tracker = {'a' : 0, 'b' : 0, 'c' : 0}
    res = 0
    start = 0
    for i in range(0,len(s)):
        tracker[s[i]] += 1
        while tracker['a'] > 0 and tracker['b'] > 0 and tracker['c'] > 0:
            res += len(s) - i
            tracker[s[start]] -= 1
            start += 1
    return res


print(num_substr("abc"))
print(num_substr("abcabc"))
print(num_substr("aaabc"))

File: leetcode/medium/substr_three_chars.cc
// Leetcode :  https://leetcode.com/problems/number-of-substrings-containing-all-three-characters/
//
//
//
//

#include <string>
#include <iostream>
#include <unordered_map>

using namespace std;

int num_substr(string s)
{
    unordered_map<char,int> tracker;
    int res = 0;
    int start = 0;
    tracker['a'] = 0;
    tracker['b'] = 0;
    tracker['c'] = 0;

    for (int i = 0; i < s.size(); ++i)
    {
        tracker[s[i]]++;

        while(tracker['a'] > 0 && tracker['b'] > 0 && tracker['c'] > 0)
        {
            res += s.size() - i;
            tracker[s[start]]--;
            start++;
        }
    }
    return res;
}

int main()
{
    cout << num_substr("abcabc") << endl;
    cout << num_substr("aaabc") << endl;
    cout << num_substr("abc") << endl;
}
